### PR TITLE
Autodetect hive_partitioning

### PR DESCRIPTION
**Pull request by lverdoes**
This PR introduces auto detection for hive_partitioning.

If the provided files appear to have the same schema and match a hive structure eg: '/year=x/month=y/file.parquet', then the `hive_partitioning` flag will be enabled automatically for that query. 

Setting `hive_partitioning=` manually will turn off auto detection.

@samansmink

### CODE REVIEW

File: src/common/multi_file_reader.cpp [hunk 1]
```diff
@@ -99,6 +100,8 @@ MultiFileReaderBindData MultiFileReader::BindOptions(MultiFileReaderOptions &opt
 	}
 
 	// Add generated constant columns from hive partitioning scheme
+	// bool hive_partitioning = options.auto_detect_hive_partitioning ? AutoDetectHivePartitioning::AutoDetect(files) :
```
Comment (samansmink) [inline src/common/multi_file_reader.cpp:103]: should this be commented out? I think the auto-detect is now disabled?

**Review by samansmink — COMMENTED** [review]
Thanks for the PR @lverdoes, looks very cool! I added a few comments

File: src/include/duckdb/common/multi_file_reader_options.hpp [hunk 2]
```diff
@@ -18,11 +20,82 @@ struct BindInfo;
...
+				}
+			}
+		}
+		return true;
+	}
+
+	static bool AutoDetectHivePartitioning(const vector<string> &files) {
+
+#ifdef __linux__
+		return AutoDetectHivePartitioningSplit(files);
```
[the remark this answers is not in the document]
Comment (lverdoes) [reply src/include/duckdb/common/multi_file_reader_options.hpp:94]: Ah, maybe the FileSystem::PathSeperator() from the above comment can solve this in 1 implementation, let's see..

**Comment by Mytherin** [comment]
Thanks for the PR! LGTM

**Pull request merged**